File: mac/Sources/LegionControlCore/AgentBundle.swift
import CryptoKit
import Foundation

/// The signed archive for upgrading an authenticated existing agent.
///
/// Three files live in `Contents/Resources/agent/`: the tarball, the signed release-schema manifest
/// that names it, and the detached signature over that manifest. They are put there by the release
/// build. A development build made without them still runs and still offers the action — disabled,
/// with the reason spelled out, because an action that quietly disappears is one nobody can ask
/// about.
struct AgentBundle: Sendable {
    var archiveURL: URL
    var manifestURL: URL
    var signatureURL: URL
    var version: String

    typealias TrustFailure = ReleaseTrust.TrustFailure

    /// The bundle this build shipped with, or nil when it shipped without one.
    static let shared: AgentBundle? = locate()

    /// The key the archive's manifest is checked against. Set only by the tests, which have to be
    /// able to produce a signed bundle of their own; the shipping app leaves it nil and the pinned
    /// release key is used.
    nonisolated(unsafe) static var verificationKey: Curve25519.Signing.PublicKey?

    /// Why there is nothing to install, for the disabled action's explanation.
    static let unavailableReason = "The signed agent archive is not included in this build."

    private static func locate(in resources: URL? = nil) -> AgentBundle? {
        let directory = resources ?? Bundle.main.resourceURL?.appending(path: "agent")
        guard let directory else { return nil }
        let manifest = directory.appending(path: ReleaseTrust.Artifacts.agentManifest)
        let signature = directory.appending(path: ReleaseTrust.Artifacts.agentManifestSignature)
        guard FileManager.default.isReadableFile(atPath: manifest.path(percentEncoded: false)),
              FileManager.default.isReadableFile(atPath: signature.path(percentEncoded: false))
        else { return nil }

        // The version comes out of the manifest rather than out of a file name, so the app never
        // claims to be shipping a version the signature does not cover.
        guard let bytes = try? Data(contentsOf: manifest),
              let signatureBytes = try? Data(contentsOf: signature),
              let verified = try? ReleaseTrust.verifiedManifest(bytes: bytes, signature: signatureBytes,
                                                                key: verificationKey),
              let version = verified.agentVersion ?? verified.artifacts.first.map({ agentVersion(from: $0.name) }) ?? nil
        else { return nil }

        let archive = directory.appending(path: ReleaseTrust.Artifacts.agentArchive(version: version))
        guard FileManager.default.isReadableFile(atPath: archive.path(percentEncoded: false)) else { return nil }
        return AgentBundle(archiveURL: archive, manifestURL: manifest, signatureURL: signature, version: version)
    }

    /// For the tests: a bundle rooted anywhere.
    static func at(_ directory: URL) -> AgentBundle? { locate(in: directory) }

    /// `legionctl-agent-3.0.0.tgz` to `3.0.0`.
    static func agentVersion(from name: String) -> String? {
        guard name.hasPrefix("legionctl-agent-"), name.hasSuffix(".tgz") else { return nil }
        return String(name.dropFirst("legionctl-agent-".count).dropLast(".tgz".count))
    }

    /// The archive bytes, checked against the signed manifest before they leave this Mac.
    ///
    /// A tampered archive is refused before a single byte is uploaded, including when upgrading
    /// a legacy agent that cannot perform this verification itself.
    func verifiedArchive() throws -> Data {
        let manifestBytes = try read(manifestURL, what: "the agent manifest")
        let signatureBytes = try read(signatureURL, what: "the agent manifest signature")
        let manifest = try ReleaseTrust.verifiedManifest(bytes: manifestBytes, signature: signatureBytes,
                                                         key: Self.verificationKey)

        let name = archiveURL.lastPathComponent
        guard let artifact = manifest.artifact(named: name) else {
            throw TrustFailure("The signed agent manifest does not name \(name), so it was not sent.")
        }
        let payload = try read(archiveURL, what: "the agent archive")
        guard payload.count == artifact.size, ReleaseTrust.sha256(of: payload) == artifact.sha256 else {
            throw TrustFailure("The agent archive does not match its signed size and hash, so it was not sent.")
        }
        return payload
    }

    private func read(_ url: URL, what: String) throws -> Data {
        guard let data = try? Data(contentsOf: url) else {
            throw TrustFailure("\(what) could not be read from this app's own resources.")
        }
        return data
    }
}

File: android/app/src/main/java/com/x1f4r/legioncontrol/agent/AgentBundle.kt
package com.x1f4r.legioncontrol.agent

import android.content.Context
import com.x1f4r.legioncontrol.data.sha256Hex
import com.x1f4r.legioncontrol.net.ManifestVerificationFailed
import com.x1f4r.legioncontrol.net.ReleaseAssets
import com.x1f4r.legioncontrol.net.ReleaseVerification
import com.x1f4r.legioncontrol.net.VerifiedManifest
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

/**
 * The signed control agent this build carries, ready to be put on a machine.
 *
 * The app bundles the agent tarball and its signed manifest from the release it came from, so that
 * putting contract 3 on a machine that has contract 2 is one action rather than a terminal session.
 * Nothing about it is trusted because it shipped inside the APK: the manifest is checked against the
 * pinned release key and the tarball against the manifest, here, before a byte of it is sent
 * anywhere. An APK can be repacked; a signature cannot be forged by repacking one.
 *
 * A build without the bundle is normal and expected: every build and every test works before a signed
 * bundle exists, and the install action is then explicitly unavailable rather than silently missing
 * or, far worse, quietly installing something unverified.
 */
class AgentBundle private constructor(
    /** The tarball, copied out of assets into cache so it has a path and a stream. */
    val tarball: File,
    val manifest: VerifiedManifest,
    val version: String,
) {
    val fileName: String get() = tarball.name

    /** What to call it on screen. */
    fun describe(): String = "control agent $version, signed"

    companion object {
        /** Where the release build puts the three files. */
        private const val ASSET_DIR = "agent"

        /**
         * Reads and verifies the bundle, or explains why there is not one.
         *
         * Every failure is a refusal. There is deliberately no path that returns a tarball which has
         * not been checked, because a caller holding one would eventually send it.
         */
        suspend fun load(context: Context): Result<AgentBundle> = withContext(Dispatchers.IO) {
            val assets = context.assets
            val names = runCatching { assets.list(ASSET_DIR)?.toList().orEmpty() }
                .getOrDefault(emptyList())
            if (names.isEmpty()) {
                return@withContext Result.failure(
                    AgentBundleMissing(
                        "This build carries no signed control agent bundle, so it cannot install " +
                            "one. A release build does; a development build does not.",
                    ),
                )
            }

            fun read(name: String): ByteArray? = runCatching {
                assets.open("$ASSET_DIR/$name").use { it.readBytes() }
            }.getOrNull()

            val manifestBytes = read(ReleaseAssets.AGENT_MANIFEST)
            val signature = read(ReleaseAssets.AGENT_MANIFEST_SIGNATURE)?.toString(Charsets.UTF_8)

            val verified = ReleaseVerification.verify(manifestBytes, signature, expectedVersion = null)
                .getOrElse { failure ->
                    return@withContext Result.failure(
                        AgentBundleMissing(
                            (failure as? ManifestVerificationFailed)?.problem?.sentence
                                ?: "the bundled agent manifest could not be verified",
                        ),
                    )
                }

            val agentVersion = verified.agentVersion.takeIf { it.isNotBlank() }
                ?: return@withContext Result.failure(
                    AgentBundleMissing("The bundled agent manifest does not say which agent version it is."),
                )
            val tarballName = ReleaseAssets.agentTarball(agentVersion)
            val artifact = verified.artifact(tarballName)
                ?: return@withContext Result.failure(
                    AgentBundleMissing("The bundled manifest does not list $tarballName."),
                )
            val bytes = read(tarballName)
                ?: return@withContext Result.failure(
                    AgentBundleMissing("This build's manifest lists $tarballName and the file is not there."),
                )

            if (artifact.size > 0 && bytes.size.toLong() != artifact.size) {
                return@withContext Result.failure(
                    AgentBundleMissing(
                        "The bundled $tarballName is ${bytes.size} bytes and its signed manifest " +
                            "says ${artifact.size}.",
                    ),
                )
            }
            if (!sha256Hex(bytes).equals(artifact.sha256, ignoreCase = true)) {
                return@withContext Result.failure(
                    AgentBundleMissing(
                        "The bundled $tarballName does not match the checksum in its signed manifest.",
                    ),
                )
            }

            // Written out only after every check has passed. A file on disk is a file something can
            // send, and one that failed verification should never exist as a path at all.
            val directory = File(context.cacheDir, "agent-bundle").apply { mkdirs() }
            directory.listFiles()?.forEach { it.delete() }
            val file = File(directory, tarballName)
            file.writeBytes(bytes)

            Result.success(AgentBundle(file, verified, agentVersion))
        }
    }
}

/** There is no usable bundle, and this is the sentence to show instead of an install action. */
class AgentBundleMissing(override val message: String) : Exception(message)

/**
 * How to get a signed agent onto one machine.
 *
 * Two paths, chosen by what is already there. A machine that already speaks contract 3 takes the
 * bytes on standard input, which needs no writable temporary path and works through the restricted
 * dispatcher. A machine on 2.x has no `self-update` at all, so the tarball goes over SFTP and the
 * staged tree installs itself, which is the bootstrap the contract describes.
 *
 * SFTP rather than `cat > path`: a redirect is shell syntax, and this app's whole quoting discipline
 * exists so that it never has to write any.
 */
enum class AgentDeploymentPath {
    /** `self-update --stdin`, streaming the tarball. Contract 3 and later. */
    STDIN,

    /** SFTP into `<base>/incoming/` then run the staged tree's own `self-update --install`. */
    UPLOAD_AND_INSTALL,
    ;

    companion object {
        fun forAgent(abilities: AgentAbilities): AgentDeploymentPath =
            if (abilities.speaksV3) STDIN else UPLOAD_AND_INSTALL
    }
}
